File: ambeth/Ambeth.Cache/ambeth/cache/rootcachevalue/RootCacheValueEnhancementHint.cs
using De.Osthus.Ambeth.Bytecode;
using System;

namespace De.Osthus.Ambeth.Cache.Rootcachevalue
{
    public class RootCacheValueEnhancementHint : IEnhancementHint, ITargetNameEnhancementHint
    {
        protected readonly Type entityType;

        public RootCacheValueEnhancementHint(Type entityType)
        {
            this.entityType = entityType;
        }

        public Type EntityType
        {
            get
            {
                return entityType;
            }
        }

        public override bool Equals(Object obj)
        {
            if (obj == this)
            {
                return true;
            }
            if (!(obj is RootCacheValueEnhancementHint))
            {
                return false;
            }
            RootCacheValueEnhancementHint other = (RootCacheValueEnhancementHint)obj;
            return EntityType.Equals(other.EntityType);
        }

        public override int GetHashCode()
        {
            return GetType().GetHashCode() ^ EntityType.GetHashCode();
        }

        public T Unwrap<T>() where T : IEnhancementHint
        {
            return (T)Unwrap(typeof(T));
        }

        public Object Unwrap(Type includedHintType)
        {
            if (typeof(RootCacheValueEnhancementHint).IsAssignableFrom(includedHintType))
            {
                return this;
            }
            return null;
        }

        public String GetTargetName(Type typeToEnhance)
        {
            return EntityType.FullName + "$" + typeof(RootCacheValue).Name;
        }
    }
}

File: jambeth/jambeth-cache/src/main/java/de/osthus/ambeth/cache/rootcachevalue/RootCacheValueEnhancementHint.java
package de.osthus.ambeth.cache.rootcachevalue;

import java.io.Serializable;

import de.osthus.ambeth.bytecode.IEnhancementHint;
import de.osthus.ambeth.bytecode.ITargetNameEnhancementHint;
import de.osthus.ambeth.repackaged.org.objectweb.asm.Type;

public class RootCacheValueEnhancementHint implements IEnhancementHint, ITargetNameEnhancementHint, Serializable
{
	private static final long serialVersionUID = 5722369699026975653L;

	protected final Class<?> entityType;

	public RootCacheValueEnhancementHint(Class<?> entityType)
	{
		this.entityType = entityType;
	}

	public Class<?> getEntityType()
	{
		return entityType;
	}

	@Override
	public boolean equals(Object obj)
	{
		if (obj == this)
		{
			return true;
		}
		if (!(obj instanceof RootCacheValueEnhancementHint))
		{
			return false;
		}
		RootCacheValueEnhancementHint other = (RootCacheValueEnhancementHint) obj;
		return getEntityType().equals(other.getEntityType());
	}

	@Override
	public int hashCode()
	{
		return RootCacheValueEnhancementHint.class.hashCode() ^ getEntityType().hashCode();
	}

	@SuppressWarnings("unchecked")
	@Override
	public <T extends IEnhancementHint> T unwrap(Class<T> includedHintType)
	{
		if (RootCacheValueEnhancementHint.class.isAssignableFrom(includedHintType))
		{
			return (T) this;
		}
		return null;
	}

	@Override
	public String toString()
	{
		return getClass().getSimpleName() + ": " + getTargetName(null);
	}

	@Override
	public String getTargetName(Class<?> typeToEnhance)
	{
		return Type.getInternalName(entityType) + "$" + RootCacheValue.class.getSimpleName();
	}
}
